openvdb/openvdb/Types.h

```diff
@@ -256,6 +264,25 @@ struct ValueTraits<T, false>
 ////////////////////////////////////////
 
 
+// @brief Extend std::is_floating_point<T>::value to work with half
+template <typename T>
+struct IsFloatingPoint { enum { value = std::is_floating_point<T>::value }; };
+#if OPENVDB_USE_HALF
+template <> struct IsFloatingPoint<half> { enum { value = true }; };
+#endif
+
+
+// @brief Extend std::is_class<T>::value to work with half
+template <typename T>
+struct IsClass { enum { value = std::is_class<T>::value }; };
+#if OPENVDB_USE_HALF
+template <> struct IsClass<half> { enum { value = false }; };
+#endif
```

Comment:
Not sure I agree with this - half is technically a class right?

Reply:
Half is a class containing one member - an unsigned short. What we're really after knowing here is if the type is a component type (ie vector, matrix, quaternion) or a simple POD-type (half, short, float, etc).

Reply:
> Half is a class containing one member

That's it then! Doesn't matter if it's a single POD member or just an interface, you'd expect the implementation of `std::is_class` to return true if the type is a class. Additionally this new struct you've introduced is called `IsClass` which you'd expect to also return true.

Taking a step back and looking at its use here:

```
template<typename ValueType_, typename Codec_>
bool
TypedAttributeArray<ValueType_, Codec_>::valueTypeIsClass() const
{
    // half is not defined as a non-class type as expected, so explicitly exclude it
    return std::is_class<ValueType>::value && !std::is_same<half, ValueType>::value;
}
```

It feels like this method is poorly named for the intended result. As you say:

> What we're really after knowing here is if the type is a component type

So perhaps it should be renamed as such. Either way, I definitely don't think we should be introducing an `IsClass` struct in this way for `half`

Reply:
Would an `IsScalar` be more appropriate?

Reply:
I think the root problem here is that I disagree with the behaviour of `TypedAttributeArray::valueTypeIsClass`. As this logic is only used there (I think) there's two solutions. The first doesn't address the issue but, for this PR, we simply remove `IsClass` and introduce the compiler branching logic in `valueTypeIsClass()` to call `!std::is_same<half, ValueType>::value` if `half` if enabled. The second and better solution would be to deprecate `valueTypeIsClass()` and introduce more appropriate methods such as `valueTypeIsPOD()` (`std::is_pod<>`) or better `valueTypeIsTrivial() valueTypeStandardLayout()` (`std::is_trivial<> std::is_standard_layout()`)

Reply:
None of these alternatives are really appropriate - std::is_pod is being deprecated in C++20, std::is_trivial and std::is_standard_layout would all return true for simple component structs (Vec, Quat, Matrix). Again, what we really want to know is if the type contains more than one (nested) value. std::is_scalar or std::is_arithmetic might work. We could change valueTypeIsClass to valueTypeIsScalar, I think I originally considered this but can't remember why I went with class in the end.

Anyway, I don't want to derail this PR as it's a bit away from the scope of this change, so I'll put this into the method itself and we can discuss changing it in a subsequent effort, which is a bit more involved as it requires phasing out through ABI.

Reply:
Yeah, it seems like the standard doesn't want people using its `type_traits` metafunctions to talk about scalar-like user-defined types: Both `std::is_floating_point` and `std::is_arithmetic` say it's UB to specialize. :-/

`openvdb/Types.h` does have `ValueTraits<T>` including `ValueTraits<T>::IsScalar`, FWIW.

Reply:
That's a very good point, in fact this may have been what motivated me to add ValueTraits. Thank you for reminding me about this one. :+1: 

I've updated this PR to convert all these methods over to use ValueTraits now.